Sources/AsyncAlgorithms/Debounce/DebounceStorage.swift

```diff
@@ -0,0 +1,296 @@
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+final class DebounceStorage<Base: AsyncSequence, C: Clock>: @unchecked Sendable where Base: Sendable {
+    typealias Element = Base.Element
+
+    /// The state machine protected with a lock.
+    private let stateMachine: ManagedCriticalState<DebounceStateMachine<Base, C>>
+    /// The interval to debounce.
+    private let interval: C.Instant.Duration
+    /// The tolerance for the clock.
+    private let tolerance: C.Instant.Duration?
+    /// The clock.
+    private let clock: C
+
+    init(base: Base, interval: C.Instant.Duration, tolerance: C.Instant.Duration?, clock: C) {
+        self.stateMachine = .init(.init(base: base, clock: clock, interval: interval))
+        self.interval = interval
+        self.tolerance = tolerance
+        self.clock = clock
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.stateMachine.withCriticalRegion { $0.iteratorDeinitialized() }
+
+        switch action {
+        case .cancelTaskAndUpstreamAndClockContinuations(
+            let task,
+            let upstreamContinuation,
+            let clockContinuation
+        ):
+            upstreamContinuation?.resume(throwing: CancellationError())
+            clockContinuation?.resume(throwing: CancellationError())
+
+            task.cancel()
+
+        case .none:
+            break
+        }
+    }
+
+    func next() async rethrows -> Element? {
+        // We need to handle cancellation here because we are creating a continuation
+        // and because we need to cancel the `Task` we created to consume the upstream
+        return try await withTaskCancellationHandler {
+            // We always suspend since we can never return an element right away
+
+            let result: Result<Element?, Error> = await withUnsafeContinuation { continuation in
+                self.stateMachine.withCriticalRegion {
+                    let action = $0.next(for: continuation)
+
+                    switch action {
+                    case .startTask(let base):
+                        self.startTask(
+                            stateMachine: &$0,
+                            base: base,
+                            downstreamContinuation: continuation
+                        )
+
+                    case .resumeUpstreamContinuation(let upstreamContinuation):
+                        // This is signalling the upstream task that is consuming the upstream
+                        // sequence to signal demand.
+                        upstreamContinuation?.resume(returning: ())
```

Comment:
hmm these continuations are being resumed inside the critical region; doesn't that mean that we could potentially get a recursion in the lock?

Reply:
Resuming continuations while hold the lock is safe. From the docs of `resume`
> After resume enqueues the task, control immediately returns to the caller. The task continues executing when its executor is able to reschedule it.

This plays out to the resumed job to be enqueued in the appropriate executor and we will continue right away with our code and drop the lock.

Reply:
hmm that means perhaps we have some work to do for some optimizations for the entire framework because I was under I guess the mistaken presumption that there was an edge case where it could be resumed such that it becomes recursive